packages/adapters-library/src/adapters/pendle/products/market/pendleMarketAdapter.ts

```diff
@@ -0,0 +1,305 @@
+import { getAddress } from 'ethers'
+import { AdaptersController } from '../../../../core/adaptersController'
+import { Chain } from '../../../../core/constants/chains'
+import {
+  CacheToFile,
+  IMetadataBuilder,
+} from '../../../../core/decorators/cacheToFile'
+
+import { CustomJsonRpcProvider } from '../../../../core/provider/CustomJsonRpcProvider'
+
+import { Helpers } from '../../../../scripts/helpers'
+import { IProtocolAdapter } from '../../../../types/IProtocolAdapter'
+import {
+  AssetType,
+  GetEventsInput,
+  GetPositionsInput,
+  GetTotalValueLockedInput,
+  MovementsByBlock,
+  PositionType,
+  ProtocolAdapterParams,
+  ProtocolDetails,
+  ProtocolPosition,
+  ProtocolTokenTvl,
+  TokenType,
+  UnwrapExchangeRate,
+  UnwrapInput,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { Protocol } from '../../../protocols'
+import { fetchAllMarkets } from '../../backend/backendSdk'
+
+import { logger } from '../../../../core/utils/logger'
+import { OraclePyYtLp__factory } from '../../contracts'
+
+type Metadata = Record<
+  string,
+  {
+    protocolToken: Erc20Metadata
+    underlyingToken: Erc20Metadata
+    marketAddress: string
+    type: 'pt' | 'yt' | 'lp' | 'sy'
+  }
+>
+
+export class PendleMarketAdapter implements IProtocolAdapter, IMetadataBuilder {
+  productId = 'market'
+  protocolId: Protocol
+  chainId: Chain
+  helpers: Helpers
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+  }
+
+  async unwrap({
+    blockNumber,
+    protocolTokenAddress,
+    tokenId,
+  }: UnwrapInput): Promise<UnwrapExchangeRate> {
+    const metadata = (await this.buildMetadata())[protocolTokenAddress]
+    const underlyingToken = metadata!.underlyingToken
+
+    const oracle = OraclePyYtLp__factory.connect(
+      '0x9a9fa8338dd5e5b2188006f1cd2ef26d921650c2',
+      this.provider,
+    )
+
+    let rate: bigint
+    switch (metadata!.type) {
```

Comment:
If possible, it'd be nice to have an explanation or a reference about:
- What is this type
- What is that 900 that is passed to some of those methods